godot_project/editor/controls/dockers/workspace_docker/simulations_docker/simulations_docker_controls/atomic_structure_model_validator.gd

```diff
@@ -126,76 +126,117 @@ func has_overlapping_atoms() -> bool:
+			continue
+		overlap_center /= current_overlap_atoms.size()
 		
-		# Relax the atoms positions.
-		# For each atom, find the nearest atom and move away. Repeat up to MAX_SHAKE_ITERATIONS
-		var move_offset: float = MAX_COVALENT_RADIUS * 0.25
-		var max_separation_distance_squared: float = pow(MAX_COVALENT_RADIUS, 2.0)
-		for iteration: int in MAX_SHAKE_ITERATIONS:
-			var atoms_were_moved: bool = false
-			for atom_id: int in atoms_positions:
-				var dir := Vector3.ONE * 99999.0
-				var min_distance := 99999.0
-
-				# Find the closest atom
-				for other_atom_id: int in atoms_positions:
-					if atom_id == other_atom_id:
-						continue
-					var diff: Vector3 = atoms_positions[atom_id] - atoms_positions[other_atom_id]
-					var distance: float = diff.length_squared()
-					if distance < min_distance:
-						dir = diff
-						min_distance = distance
-				
-				if min_distance > max_separation_distance_squared:
-					continue # Atom already far enough
-				
-				atoms_positions[atom_id] += dir.normalized() * move_offset
-				atoms_were_moved = true
+		# Move each overlapping atoms away from the center point, by at least
+		# their atomic radius.
+		for atom_id: int in current_overlap_atoms:
+			var item: SpatialHashGrid.Item = current_overlap_atoms[atom_id]
+			var position: Vector3 = item.position
+			if position.is_equal_approx(overlap_center):
+				position += Vector3(randf() - 0.5, randf() - 0.5, randf() - 0.5) * 0.1
+			var dir_from_center: Vector3 = overlap_center.direction_to(position)
+			var atomic_number: int = nano_structure.atom_get_atomic_number(atom_id)
+			var element_data: ElementData = PeriodicTable.get_by_atomic_number(atomic_number)
+			var radius: float = element_data.covalent_radius[1]
+			const DISTANCE_MARGIN: float = 1.1
+			var new_position: Vector3 = overlap_center + dir_from_center * radius * DISTANCE_MARGIN
+			overlap_atom_grid.move_item(item, new_position)
+	
+	# Atoms might still overlap with other overlap groups, or with the rest of
+	# the structure (described in static_atom_grid).
+	# Relax all the overlapping atoms in the same pass.
+	# For each atom, find the nearest atom and move away. Repeat up to MAX_SHAKE_ITERATIONS
+	var multiplier: float = 1.0
+	for iteration: int in MAX_SHAKE_ITERATIONS:
+		var atoms_were_moved: bool = false
+		for item: SpatialHashGrid.Item in overlap_atom_grid.get_all_items():
+			# Find the closest atom
+			var nearby_atoms: Array[SpatialHashGrid.Item] = []
+			nearby_atoms = overlap_atom_grid.get_items_around(item.position)
+			nearby_atoms.append_array(static_atom_grid.get_items_around(item.position))
+			var closest_atom: SpatialHashGrid.Item
+			var min_distance: float = INF
+			var separation_dir: Vector3
+			for candidate_item in nearby_atoms:
+				if item == candidate_item:
+					continue
+				var diff: Vector3 = item.position - candidate_item.position
+				var distance: float = diff.length_squared()
+				if distance < min_distance:
+					separation_dir = diff
```

Comment:
nit: because this variable claims to be a direction, better to normalize here instead of line 214

Reply:
I've renamed the variable instead so the normalized() call only happens once